pkg/collector/corechecks/containers/containerd.go

```diff
@@ -0,0 +1,359 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build containerd
+
+package containers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/containerd/cgroups"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/typeurl"
+	"github.com/gogo/protobuf/types"
+	"gopkg.in/yaml.v2"
+
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	cutil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	containerdCheckName = "containerd"
+)
+
+// ContainerCheck grabs containerd metrics and events
+type ContainerdCheck struct {
+	core.CheckBase
+	hostname string
+	instance *ContainerdConfig
+	sub      *Subscriber
+}
+
+// ContainerdConfig contains the custom options and configurations set by the user.
+type ContainerdConfig struct {
+	Tags    []string `yaml:"tags"`
+	Filters []string `yaml:"filters"`
+}
+
+func init() {
+	corechecks.RegisterCheck(containerdCheckName, ContainerdFactory)
+}
+
+// ContainerdFactory is used to create register the check and initialize it.
+func ContainerdFactory() check.Check {
+	return &ContainerdCheck{
+		CheckBase: corechecks.NewCheckBase(containerdCheckName),
+		instance:  &ContainerdConfig{},
+		sub:       &Subscriber{},
+	}
+}
+
+// Parse is used to get the configuration set by the user
+func (co *ContainerdConfig) Parse(data []byte) error {
+	if err := yaml.Unmarshal(data, co); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Configure parses the check configuration and init the check
+func (c *ContainerdCheck) Configure(config, initConfig integration.Data) error {
+	var err error
+	if err = c.CommonConfigure(config); err != nil {
+		return err
+	}
+
+	c.hostname, err = util.GetHostname()
+	if err != nil {
+		return err
+	}
+
+	if err = c.instance.Parse(config); err != nil {
+		return err
+	}
+	c.sub.Filters = c.instance.Filters
+	return nil
+}
+
+// Run executes the check
+func (c *ContainerdCheck) Run() error {
+	sender, err := aggregator.GetSender(c.ID())
+	defer sender.Commit()
```

Comment:
if `err != nil`, `sender` might be `nil`, we should move the `defer sender.Commit()` after the error check